kv/kvserver: TestGossipHandlesReplacedNode failed

kv/kvserver.TestGossipHandlesReplacedNode [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4044053&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4044053&tab=artifacts#/) on master @ [260144b9bec2b2da2cf96c7198197a88b97c942c](https://github.com/cockroachdb/cockroach/commits/260144b9bec2b2da2cf96c7198197a88b97c942c):


```
=== RUN   TestGossipHandlesReplacedNode
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestGossipHandlesReplacedNode3951923715
    test_log_scope.go:80: use -show-logs to present logs inline
    testcluster.go:412: listen tcp 127.8.243.194:36669: bind: address already in use
    panic.go:642: -- test log scope end --
--- FAIL: TestGossipHandlesReplacedNode (1.81s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGossipHandlesReplacedNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-12084

kv/kvserver.TestGossipHandlesReplacedNode [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4634062&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4634062&tab=artifacts#/) on master @ [54671e7befc13ef0d4885c8506831bb0c6182ffc](https://github.com/cockroachdb/cockroach/commits/54671e7befc13ef0d4885c8506831bb0c6182ffc):


```
=== RUN   TestGossipHandlesReplacedNode
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestGossipHandlesReplacedNode2038960156
    test_log_scope.go:80: use -show-logs to present logs inline
    testcluster.go:417: listen tcp 127.103.61.243:43155: bind: address already in use
    panic.go:642: -- test log scope end --
--- FAIL: TestGossipHandlesReplacedNode (1.95s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGossipHandlesReplacedNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/gossip_test.go
+++ b/pkg/kv/kvserver/gossip_test.go
@@ -27,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/dbdesc"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util"
@@ -183,7 +184,17 @@ func TestGossipHandlesReplacedNode(t *testing.T) {
 	newServerArgs.JoinAddr = tc.Servers[1].ServingRPCAddr()
 	log.Infof(ctx, "stopping server %d", oldNodeIdx)
 	tc.StopServer(oldNodeIdx)
-	tc.AddAndStartServer(t, newServerArgs)
+	// We are re-using a hard-coded port. Other processes on the system may by now
+	// be listening on this port, so there will be flakes. For now, skip the test
+	// when this flake occurs.
+	//
+	// The real solution would be to create listeners for both RPC and SQL at the
+	// beginning of the test, and to make sure they aren't closed on server
+	// shutdown. Then we can pass the listeners to the second invocation. Alas,
+	// this requires some refactoring that remains out of scope for now.
+	if err := tc.AddAndStartServerE(newServerArgs); err != nil && !testutils.IsError(err, `address already in use`) {
+		t.Fatal(err)
+	}
 
 	tc.WaitForNStores(t, tc.NumServers(), tc.Server(1).GossipI().(*gossip.Gossip))
 

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -399,23 +399,27 @@ func checkServerArgsForCluster(
 	return nil
 }
 
-// AddAndStartServer creates a server with the specified arguments and appends it to
+// AddAndStartServer calls through to AddAndStartServerE.
+func (tc *TestCluster) AddAndStartServer(t *testing.T, serverArgs base.TestServerArgs) {
+	t.Helper()
+	require.NoError(t, tc.AddAndStartServerE(serverArgs))
+}
+
+// AddAndStartServerE creates a server with the specified arguments and appends it to
 // the TestCluster. It also starts it.
 //
 // The new Server's copy of serverArgs might be changed according to the
 // cluster's ReplicationMode.
-func (tc *TestCluster) AddAndStartServer(t testing.TB, serverArgs base.TestServerArgs) {
+func (tc *TestCluster) AddAndStartServerE(serverArgs base.TestServerArgs) error {
 	if serverArgs.JoinAddr == "" && len(tc.Servers) > 0 {
 		serverArgs.JoinAddr = tc.Servers[0].ServingRPCAddr()
 	}
 	_, err := tc.AddServer(serverArgs)
 	if err != nil {
-		t.Fatal(err)
+		return err
 	}
 
-	if err := tc.startServer(len(tc.Servers)-1, serverArgs); err != nil {
-		t.Fatal(err)
-	}
+	return tc.startServer(len(tc.Servers)-1, serverArgs)
 }
 
 // AddServer is like AddAndStartServer, except it does not start it.

